lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationProviderServiceImpl.java

```diff
@@ -28,190 +31,320 @@
+      fusionEngineFactory.setContext(context);
+      fusionEngineFactory.setCallback(this);
+    }
+    locationEngineFactory = engineFactory;
+    clientManager.setListener(clientManagerListener);
     listenerToRequest = new HashMap<>();
     intentToRequest = new HashMap<>();
     callbackToRequest = new HashMap<>();
     callbackToLooper = new HashMap<>();
+    clientToLocationEngine = new HashMap<>();
+    clientToMockMode = new HashMap<>();
   }
 
   public void shutdown() {
+    for (LocationEngine engine : clientToLocationEngine.values()) {
+      engine.setRequest(null);
+    }
     listenerToRequest.clear();
     intentToRequest.clear();
     callbackToRequest.clear();
     callbackToLooper.clear();
-    locationEngine.setRequest(null);
+    clientToLocationEngine.clear();
+    clientToMockMode.clear();
   }
 
-  public Location getLastLocation() {
+  public Location getLastLocation(LostApiClient apiClient) {
+    LocationEngine locationEngine = clientToLocationEngine.get(apiClient);
     return locationEngine.getLastLocation();
   }
 
-  public LocationAvailability getLocationAvailability() {
-    return createLocationAvailability();
+  @RequiresPermission(anyOf = {ACCESS_COARSE_LOCATION, ACCESS_FINE_LOCATION})
+  public LocationAvailability getLocationAvailability(LostApiClient apiClient) {
+    LocationEngine locationEngine = clientToLocationEngine.get(apiClient);
+    return locationEngine.createLocationAvailability();
   }
 
-  public void requestLocationUpdates(LocationRequest request, LocationListener listener) {
-    listenerToRequest.put(listener, request);
-    locationEngine.setRequest(request);
-  }
+  public void requestLocationUpdates(LostApiClient apiClient, LocationRequest request,
+      LocationListener listener) {
+    Map<LocationListener, LocationRequest> map = listenerToRequest.get(apiClient);
+    if (map == null) {
+      map = new HashMap<>();
+    }
+    map.put(listener, request);
+    listenerToRequest.put(apiClient, map);
 
-  public void requestLocationUpdates(LocationRequest request, PendingIntent callbackIntent) {
-    intentToRequest.put(callbackIntent, request);
+    LocationEngine locationEngine = clientToLocationEngine.get(apiClient);
     locationEngine.setRequest(request);
   }
 
-  public void requestLocationUpdates(LocationRequest request, LocationCallback callback,
-      Looper looper) {
-    callbackToRequest.put(callback, request);
-    callbackToLooper.put(callback, looper);
+  public void requestLocationUpdates(LostApiClient apiClient, LocationRequest request,
+      PendingIntent callbackIntent) {
+    Map<PendingIntent, LocationRequest> map = intentToRequest.get(apiClient);
+    if (map == null) {
+      map = new HashMap<>();
+    }
+    map.put(callbackIntent, request);
+    intentToRequest.put(apiClient, map);
+
+    LocationEngine locationEngine = clientToLocationEngine.get(apiClient);
     locationEngine.setRequest(request);
```

Comment:
The fused location sample activity is crashing on startup because`clientToLocationEngine.get(apiClient)` returns `null` for the location engine. It appears when the engine is created it does not get properly registered.

So that brings up two questions:

1. How can we ensure clients and location engines are properly registered?
2. How should we handle the error case when a request is made using client that is not registered?

```java
08-31 11:52:24.935 29271 29271 E AndroidRuntime: FATAL EXCEPTION: main
08-31 11:52:24.935 29271 29271 E AndroidRuntime: Process: com.example.lost, PID: 29271
08-31 11:52:24.935 29271 29271 E AndroidRuntime: java.lang.NullPointerException: Attempt to invoke virtual method 'void com.mapzen.android.lost.internal.LocationEngine.setRequest(com.mapzen.android.lost.api.LocationRequest)' on a null object reference
08-31 11:52:24.935 29271 29271 E AndroidRuntime:        at com.mapzen.android.lost.internal.FusedLocationProviderServiceImpl.requestLocationUpdates(FusedLocationProviderServiceImpl.java:103)
08-31 11:52:24.935 29271 29271 E AndroidRuntime:        at com.mapzen.android.lost.internal.FusedLocationProviderService.requestLocationUpdates(FusedLocationProviderService.java:72)
08-31 11:52:24.935 29271 29271 E AndroidRuntime:        at com.mapzen.android.lost.internal.FusedLocationProviderApiImpl.requestLocationUpdates(FusedLocationProviderApiImpl.java:94)
08-31 11:52:24.935 29271 29271 E AndroidRuntime:        at com.example.lost.FusedLocationApiActivity.onConnected(FusedLocationApiActivity.java:173)
08-31 11:52:24.935 29271 29271 E AndroidRuntime:        at com.mapzen.android.lost.internal.FusedLocationProviderApiImpl$1.onServiceConnected(FusedLocationProviderApiImpl.java:43)
08-31 11:52:24.935 29271 29271 E AndroidRuntime:        at android.app.LoadedApk$ServiceDispatcher.doConnected(LoadedApk.java:1223)
08-31 11:52:24.935 29271 29271 E AndroidRuntime:        at android.app.LoadedApk$ServiceDispatcher$RunConnection.run(LoadedApk.java:1240)
08-31 11:52:24.935 29271 29271 E AndroidRuntime:        at android.os.Handler.handleCallback(Handler.java:739)
08-31 11:52:24.935 29271 29271 E AndroidRuntime:        at android.os.Handler.dispatchMessage(Handler.java:95)
08-31 11:52:24.935 29271 29271 E AndroidRuntime:        at android.os.Looper.loop(Looper.java:148)
08-31 11:52:24.935 29271 29271 E AndroidRuntime:        at android.app.ActivityThread.main(ActivityThread.java:5417)
08-31 11:52:24.935 29271 29271 E AndroidRuntime:        at java.lang.reflect.Method.invoke(Native Method)
08-31 11:52:24.935 29271 29271 E AndroidRuntime:        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
08-31 11:52:24.935 29271 29271 E AndroidRuntime:        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
08-31 11:52:24.937  1182  2999 W ActivityManager:   Force finishing activity com.example.lost/.FusedLocationApiActivity
08-31 11:52:24.943  1182  2999 W ActivityManager:   Force finishing activity com.example.lost/.MainActivity
```